### ISSUE TITLE
Use JSON file to check for new releases (instead of RPC server)

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#839]
The new package.json file (that will reside in both GitHub repo and in root of Textpattern.com) will be used to check if new versions are available, instead of the old RPC server (which we are gradually moving away from).

We propose checking the Textpattern.com root first, then if unavailable check GitHub repo as a fallback.

**Comment by rwetzlmayr** [comment]
How will a maintainer deploy textpattern.com/package.json?

I suggest to take a look at the current implementation, i.e. Kusor's `txp_current_version` plugin the at the RPC server as it allows maintaining this information from the Textpattern admin area without needing any other server access methods (ssh et al.)

**Comment by philwareham** [comment]
They would need FTP access to .com to update the JSON file, I guess. Other option is to serve the JSON just from GitHub but:

a) That relies on a third party.
b) It may break GitHubs terms of use

If the `txp_current_version` plugin is something simple that can run on .com without modifications to Textpattern core, then I'm interested. It would need some development so that we can provide a stable release number, as well as separate prerelease number when beta versions are available. The diagnostic panel can then give people the option of using a beta when available, or stick to stable releases.

**Comment by rwetzlmayr** [comment]
> If the txp_current_version plugin is something simple that can run on .com without modifications to Textpattern core

It is. You can look at the code on `rpc.textpattern.com/textpattern/index.php?event=plugin&step=plugin_edit&name=txp_current_version`.

**Comment by philwareham** [comment]
OK, great, I'll investigate.

Basically the major reason for having the GitHub fallback is if the .com domain ever disappears (which we all know is a possibility due to ownership). So even if we use the plugin on .com, I'd still like to fallback to GitHub. Otherwise in the (hopefully unlikely) event that .com fails - users could still be notified of an upgrade.

**Comment by rwetzlmayr** [comment]
I think and hope that the actual discovery algorithm (how a client server finds their `package.json` on the web) is not related to the method used to generate this resource on textpattern.com (or github.com, or wherever).

**Comment by philwareham** [comment]
Not wishing to pass the buck, but @Bloke will be wrong on this, so he is best placed to answer your concerns. I just opened the issue based on discussions we had at the weekend.

**Comment by rwetzlmayr** [comment]
I just discussed the issue based on it being open ;)

**Comment by petecooper** [comment]
Aside: a Textpattern page+form or page+form+article combination on textpattern.com would achieve the same effect as a proper file.

**Comment by philwareham** [comment]
@petecooper yes but it would involve a query if the database whereas a flat file wouldn't.

**Comment by petecooper** [comment]
@philwareham True. I was thinking Plan B, should FTP access not be doable.

(though practically speaking if FTP access disappears then we might as well burn the place to the ground as the last straw).

**Comment by philwareham** [comment]
OK, I now have a tidy solution that works directly from our new Textpattern.com website via the CMS - no FTP access needed or anything like that. See https://github.com/textpattern/textpattern-com-website/commit/4aeb9ae39bdf8786174033973f376b34cd16e6ad.

**Comment by philwareham** [comment]
Now the new site is live, the version check JSON file is available: https://textpattern.com/version.json

**Comment by petecooper** [comment]
Brilliant!

**Comment by philwareham** [comment]
Just a note about how it works - the JSON displays the version numbers (titles) from the latest files that are categorised as 'current release' and (if available) 'current beta release' within the CMS, using rah_external_output plugin to generate a the JSON file. Therefore only standard CMS access is needed (no FTPing or whatnot).

**Comment by bloatware** [comment]
FWIW, in 4.7 you can set headers in core with
```
<txp:php>header('Content-Type: application/json; charset=utf-8');</txp:php>
```
and so on.

**Comment by philwareham** [comment]
So we could somehow remove the need for rah_external_output to serve up our JSON file? If so can you give me some direction?

**Comment by bloatware** [comment]
Yes, but it's not very practical atm. I guess, you would need to create a form and a page `json` with something like
```
<txp:php>header('Content-Type: application/json; charset=utf-8');</txp:php>
<!-- other headers: Cache-Control etc -->
<txp:article form="json" />
```
inside. Then create a `json` section and assign to it the articles you want to output in JSON format. The combination of the form `json` and a `json` article should produce a valid JSON output in some way.

The main inconvenience is that currently txp needs a section and a page to output an article. We should think on removing this constraint.

**Comment by philwareham** [comment]
OK, thanks. I'll stick with the current method for now. Agree that is would be mighty useful in future if articles could be decoupled from sections if desired.

**Comment by bloatware** [comment]
It seems fairy easy (and logical) to consider `file_download/fileid` scheme as `file_download/form` if `fileid` is not numeric. We could then (after some validity check) output the parsed `form` à la `rah_eo`. For security sake, we would need a new `external` form type, though.

**Comment by philwareham** [comment]
Just bumping this - the `version.json` file is [up and running](https://textpattern.com/version.json) so I'm hoping the version checking code in Textpattern 4.7-dev can be changed to check this file now instead of the old RPC server? Anyone up for doing that?

**Comment by makss** [comment]
Oh sure. I'll make a patch.

**Comment by philwareham** [comment]
@makss Hi. I've [put a fake version.json file up](https://textpattern.io/version.json) with new version numbers in - but I'm not seeing any reports in Textpattern 4.7.0-dev that a new version is available. Can you confirm this is working or not? Cheers!

**Comment by makss** [comment]
Version checking does not occur in `dev`. [Set](https://github.com/textpattern/textpattern/blob/adec4dcf26eddb3f9953a04da2041ce44e7bfbb0/textpattern/index.php#L69) :
```
$txp_using_svn = $txp_is_dev = false;
```

**Comment by philwareham** [comment]
Hi again @makss - OK, I have verified that the `release` string in version.json file triggers the update message. That works great thanks.

However we do also need to check the `release` string in that version.json file too and trigger an additional message if that is not `null` and the string is a higher version than was it currently installed.

So, in simple terms, if you are currently on `4.7.0` and a prerelease `4.7..1-beta` we need a message to let people know that - I'll have to create a new string for that in Textpacks and a new pophelp item. Stating betas are used at own risk as not guaranteed to be bug free or words to that effect.

Similarly, if a user is on `4.7.0` and a `4.7.1` release is available *and* and prerelease `4.7.2-beta.3` or whatever is also available, we need to show both messages.

Sorry, hope that makes sense!

**Comment by philwareham** [comment]
Done by @makss - closing.
